### PR TITLE
Initial implementation of asynchronous requests

### LINKED ISSUE #23: Support asynchronous request handling?
Hi, I was looking at this crate library, considering using it for a project of mine, but I noticed the RPC request are not handled asynchronously, is that correct? That is, only one RPC request can be handled at a time.

### PR DESCRIPTION
**Pull request by tomusdrw**
Closes #23 

Still a bit work in progress:
- [x] Need to test integration with `jsonrpc-http-server`, `json-ipc-server`, `json-tcp-server` and `parity`.
- [x] Documentation
- [x] Examples

### CODE REVIEW

File: src/async.rs [hunk 1]
```diff
@@ -24,98 +21,110 @@ impl AsyncResult {
 	/// Callback is invoked right away if result is instantly available and `true` is returned.
 	/// `false` is returned when listener has been added
 	pub fn on_result<F>(self, f: F) -> bool where F: Fn(Res) + Send + 'static {
-		let result = self.result.clone();
-		self.result.on_result(move || {
-			f(result.await())
+		self.result.lock().unwrap().on_result(move |res| {
```
Comment (rphmeier) [inline src/async.rs]: ought to just be `on_result(f)`?

File: src/async.rs [hunk 2]
```diff
@@ -24,98 +21,110 @@ impl AsyncResult {
...
 
-
+impl Into<MethodResult> for AsyncResult {
+	fn into(self) -> MethodResult {
+		let result = self.result.lock().unwrap().try_result();
+		if let Some(result) = result {
+			MethodResult::Sync(result)
+		} else {
+			MethodResult::Async(self)
+		}
```
Comment (rphmeier) [inline src/async.rs]: ought to be a match here for conciseness.
a descriptive method name would also help readability when it's invoked.

Comment (debris) [inline src/async.rs]: imo `Into` is ok. Although I would implement `From` instead. From [rust docs](https://doc.rust-lang.org/std/convert/trait.Into.html#tymethod.into): 

> Library authors should not directly implement this trait, but should prefer implementing the `From` trait, which offers greater flexibility and provides an equivalent `Into` implementation for free, thanks to a blanket implementation in the standard library.

File: src/async.rs [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+	}
+}
+
+
+#[derive(Default)]
+struct AsyncResultHandler {
+	result: Option<Res>,
+	listener: Option<Box<FnMut(Res) + Send>>,
+	notifier: Option<Box<FnMut() + Send>>,
+}
```
Comment (debris) [inline src/async.rs]: `enum` might be a better choice here

File: src/async.rs [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+
+impl fmt::Debug for AsyncResultHandler {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+		write!(
+			formatter,
+			"AsyncResult: {:?}",
+			self.result,
```
Comment (debris) [inline src/async.rs]: could be 1 line :)

File: src/io.rs [hunk 5]
```diff
@@ -130,19 +196,38 @@ impl IoHandler {
...
+	/// Handle given request synchronously - will block until response is available.
+	pub fn handle_request_sync<'a>(&self, request_str: &'a str) -> Option<String> {
 		trace!(target: "rpc", "Request: {}", request_str);
 		let response = match read_request(request_str) {
-			Ok(request) => self.request_handler.handle_request(request).map(write_response),
-			Err(error) => Some(write_response(Response::Single(Output::Failure(Failure {
+			Ok(request) => match self.request_handler.handle_request(request) {
+				Some(response) => Some(write_response(response.await())),
+				_ => None,
+			},
```
Comment (debris) [inline src/io.rs]: `.map` might be better here

**Comment by debris** [comment]
mustntgrumble ;)

**Comment by tomusdrw** [comment]
Bump for review.

I think additional refactoring to get rid of `Mutex` in `AsyncResult` is possible, but would require significant API changes - we can aim for this in `jsonrpc-core@4.0` :)

**Pull request merged**